File: C++/2-Medium/00208._实现_Trie_(前缀树)/00208._实现_Trie_(前缀树).cpp
#include <iostream>
#include <vector>
using namespace std;

class TrieNode {
public:
    bool isEnd;
    vector<TrieNode*> children;
    TrieNode() : isEnd(false), children(26, nullptr) {}
};

class Trie {
private:
    TrieNode* root;
public:
    Trie() {
        root = new TrieNode();
    }
    
    // 插入字符串到 Trie
    void insert(string word) {
        TrieNode* node = root;
        for (char c : word) {
            int index = c - 'a';
            if (node->children[index] == nullptr) {
                node->children[index] = new TrieNode();
            }
            node = node->children[index];
        }
        node->isEnd = true;
    }
    
    // 搜索字符串在 Trie 中是否存在
    bool search(string word) {
        TrieNode* node = root;
        for (char c : word) {
            int index = c - 'a';
            if (node->children[index] == nullptr) {
                return false;
            }
            node = node->children[index];
        }
        return node->isEnd;
    }
    
    // 检查是否存在以 prefix 开头的字符串
    bool startsWith(string prefix) {
        TrieNode* node = root;
        for (char c : prefix) {
            int index = c - 'a';
            if (node->children[index] == nullptr) {
                return false;
            }
            node = node->children[index];
        }
        return true;
    }
};

int main() {
    Trie trie;
    trie.insert("apple");
    cout << trie.search("apple") << endl;   // 输出：1
    cout << trie.search("app") << endl;     // 输出：0
    cout << trie.startsWith("app") << endl; // 输出：1
    trie.insert("app");
    cout << trie.search("app") << endl;     // 输出：1
    return 0;
}

File: Python/2-Medium/00208._实现_Trie_(前缀树)/00208._实现_Trie_(前缀树).py
from typing import Dict, Optional

class TrieNode:
    """定义 Trie 节点类"""
    def __init__(self):
        # 每个节点存储对子节点的引用
        self.children: Dict[str, 'TrieNode'] = {}
        # 标记该节点是否是某个单词的结束
        self.is_end_of_word: bool = False

class Trie:
    """实现 Trie (前缀树)"""
    
    def __init__(self):
        """初始化 Trie"""
        self.root = TrieNode()
        
    def insert(self, word: str) -> None:
        """向 Trie 中插入一个单词"""
        node = self.root
        for char in word:
            # 如果字符不存在，则创建一个新的 TrieNode
            if char not in node.children:
                node.children[char] = TrieNode()
            # 移动到下一个节点
            node = node.children[char]
        # 标记单词的结束节点
        node.is_end_of_word = True
        
    def search(self, word: str) -> bool:
        """在 Trie 中搜索一个单词"""
        node = self.root
        for char in word:
            # 如果字符不在子节点中，返回 False
            if char not in node.children:
                return False
            # 移动到下一个节点
            node = node.children[char]
        # 返回最后节点的 is_end_of_word 值
        return node.is_end_of_word
        
    def startsWith(self, prefix: str) -> bool:
        """检查 Trie 中是否有以指定前缀开始的单词"""
        node = self.root
        for char in prefix:
            # 如果字符不在子节点中，返回 False
            if char not in node.children:
                return False
            # 移动到下一个节点
            node = node.children[char]
        # 如果能走到这一步，说明找到了对应的前缀
        return True

# 测试代码
trie = Trie()
trie.insert("apple")
print(trie.search("apple"))  # True
print(trie.search("app"))    # False
print(trie.startsWith("app"))# True
trie.insert("app")
print(trie.search("app"))    # True